### PR TITLE
Serve static files

### PR DESCRIPTION
**Pull request by kfrankc**
This adds support for serving up files statically (without dynamic modification like compression, code execution, or templating). HTML, PNG, and GIF files should all load in a web browser without issue.

### CODE REVIEW

File: request_handlers.cc [hunk 1]
```diff
@@ -30,14 +36,118 @@ void EchoRequestHandler::handleRequest(const char* request_buffer,
...
+  boost::tokenizer<boost::char_separator<char>> tokens(request, separator);
+  std::cout << "DEBUG: Request tokens in StaticRequestHandler::handleRequest()" << std::endl;
+  for (const auto& t : tokens) {
+    std::cout << t << "." << std::endl;
+  }
+
+  int i = 0;
+  auto curToken = tokens.begin();
+  std::string resourcePath;
+  for (;;) {
```
Comment (apfong) [inline request_handlers.cc]: Nice hack

File: request_handlers.cc [hunk 2]
```diff
@@ -30,14 +36,118 @@ void EchoRequestHandler::handleRequest(const char* request_buffer,
...
+    // We assume the 2nd token is the resource path: GET /path/to/resource.txt
+    if (i == 1) {
+      resourcePath = *curToken;
+    }
+
+    i++;
+    curToken++;
+  }
+
+  std::string request_path = resourcePath;
```
Comment (apfong) [inline request_handlers.cc]: We can probably put these two blocks together (74-90) and cite once

File: request_handlers_test.cc [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 class RequestHandlersTest : public ::testing::Test {
 protected:
+
   const size_t header_size = 45;
-  bool HandleRequest(const char* request_buffer,
-                     const size_t& request_buffer_size,
-                     char* &response_buffer,
-                     size_t& response_buffer_size) {
+
+  bool EchoHandleRequest(const char* request_buffer,
```
Comment (apfong) [inline request_handlers_test.cc]: Do we want to explicitly call certain request handlers? Not sure if it was intentional or not, but in the previous commit, I had changed them to use the HandleRequest interface.

File: request_handlers_test.cc [hunk 4]
```diff
@@ -36,10 +59,10 @@ TEST_F(RequestHandlersTest, EmptyStringTest) {
     "HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\n\r\n";
   const size_t expected_response_buffer_size = header_size + request_buffer_size;
 
-  ASSERT_TRUE(HandleRequest(request_buffer,
-                            request_buffer_size,
-                            response_buffer,
-                            response_buffer_size));
+  ASSERT_TRUE(EchoHandleRequest(request_buffer,
```
Comment (apfong) [inline request_handlers_test.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/42/files#r100144068)

File: request_handlers_test.cc [hunk 5]
```diff
@@ -55,14 +78,14 @@ TEST_F(RequestHandlersTest, RandomStringTest) {
...
-  char expected_response_buffer[] =
+  const char expected_response_buffer[] =
     "HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\n\r\nabcd\n";
   const size_t expected_response_buffer_size = header_size + request_buffer_size;
 
-  ASSERT_TRUE(HandleRequest(request_buffer,
-                            request_buffer_size,
-                            response_buffer,
-                            response_buffer_size));
+  ASSERT_TRUE(EchoHandleRequest(request_buffer,
```
Comment (apfong) [inline request_handlers_test.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/42/files#r100144068)

File: request_handlers_test.cc [hunk 6]
```diff
@@ -96,13 +119,89 @@ TEST_F(RequestHandlersTest, OutputWithHeader) {
...
+
+  // TODO: Need a way to diff two HTML files
+  // TODO: Need to convert string to char array to diff
+  std::ifstream in("static_test_files/index.html");
+  std::stringstream buffer;
+  buffer << in.rdbuf();
+  std::string contents(buffer.str());
+  std::cout << contents << std::endl;
+
+  ASSERT_TRUE(StaticHandleRequest(request_buffer,
```
Comment (apfong) [inline request_handlers_test.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/42/files#r100144068)

File: request_handlers_test.cc [hunk 7]
```diff
@@ -96,13 +119,89 @@ TEST_F(RequestHandlersTest, OutputWithHeader) {
...
+  // TODO: Need a way to diff two images; most online resources
+  // say to use OpenCV
+  // TODO: Need to convert string to char array to diff
+  std::ifstream fin("static_test_files/angrybird.png", std::ios::binary);
+  std::ostringstream buffer;
+  buffer << fin.rdbuf();
+  std::string contents(buffer.str());
+  std::cout << contents << std::endl;
+
+  ASSERT_TRUE(StaticHandleRequest(request_buffer,
```
Comment (apfong) [inline request_handlers_test.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/42/files#r100144068)

File: request_handlers_test.cc [hunk 8]
```diff
@@ -96,13 +119,89 @@ TEST_F(RequestHandlersTest, OutputWithHeader) {
...
+  // The response_buffer is allocated by the request-handler
+  char* response_buffer = nullptr;
+  size_t response_buffer_size = 0;
+
+  const char expected_response_buffer[] =
+    "HTTP/1.1 200 OK\r\nContent-Type: image/gif\r\n\r\n";
+  const size_t expected_response_buffer_size = header_size + request_buffer_size;
+
+  // TODO: Need a way to diff two GIFs
+  ASSERT_TRUE(StaticHandleRequest(request_buffer,
```
Comment (apfong) [inline request_handlers_test.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/42/files#r100144068)

File: request_handlers_test.cc [hunk 9]
```diff
@@ -96,13 +119,89 @@ TEST_F(RequestHandlersTest, OutputWithHeader) {
     "Accept-Encoding: gzip, deflate\r\nConnection: keep-alive\r\n\r\n";
   const size_t expected_response_buffer_size = header_size + request_buffer_size;
 
-  ASSERT_TRUE(HandleRequest(request_buffer,
-                            request_buffer_size,
-                            response_buffer,
-                            response_buffer_size));
+  ASSERT_TRUE(EchoHandleRequest(request_buffer,
```
Comment (apfong) [inline request_handlers_test.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/42/files#r100144068)

File: request_handlers_test.cc [hunk 10]
```diff
@@ -96,13 +119,89 @@ TEST_F(RequestHandlersTest, OutputWithHeader) {
...
+  // The request_buffer is normally filled by a socket-read
+  // The response_buffer is allocated by the request-handler
+  char* response_buffer = nullptr;
+  size_t response_buffer_size = 0;
+
+  const char expected_response_buffer[] =
+    "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n";
+  const size_t expected_response_buffer_size = header_size + request_buffer_size;
+
+  // TODO: Need a way to diff two HTML files
```
Comment (apfong) [inline request_handlers_test.cc]: Is this still a TODO?

File: request_handlers_test.cc [hunk 11]
```diff
@@ -96,13 +119,89 @@ TEST_F(RequestHandlersTest, OutputWithHeader) {
...
+  // The request_buffer is normally filled by a socket-read
+  // The response_buffer is allocated by the request-handler
+  char* response_buffer = nullptr;
+  size_t response_buffer_size = 0;
+
+  const char expected_response_buffer[] =
+    "HTTP/1.1 200 OK\r\nContent-Type: image/png\r\n\r\n";
+  const size_t expected_response_buffer_size = header_size + request_buffer_size;
+
+  // TODO: Need a way to diff two images; most online resources
```
Comment (apfong) [inline request_handlers_test.cc]: Is this still a TODO?

File: request_router.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
+#ifndef REQUEST_ROUTER_H
+#define REQUEST_ROUTER_H
+
+#include "server_config.h"
+#include <cstddef>
+
+class RequestRouter {
+  public:
+    // TODO: Reduce duplication between this and the handlers
```
Comment (apfong) [inline request_router.h]: What is this referring to?

File: request_handlers_test.cc (hunk 3 above)
Comment (KyCodeHuynh) [inline request_handlers_test.cc]: We definitely want to use the HandleRequest + RequestRouter interface. This is addressed by PR #43.

File: request_handlers_test.cc (hunk 4 above)
Comment (KyCodeHuynh) [inline request_handlers_test.cc]: We definitely want to use the HandleRequest + RequestRouter interface. This is addressed by PR #43.

File: request_handlers_test.cc (hunk 10 above)
Comment (KyCodeHuynh) [inline request_handlers_test.cc]: This is addressed by PR #43.

File: request_handlers_test.cc (hunk 11 above)
Comment (KyCodeHuynh) [inline request_handlers_test.cc]: This is addressed by PR #43.

File: request_router.h (hunk 12 above)
Comment (KyCodeHuynh) [inline request_router.h]: RequestRouter is currently just a wrapper atop the handlers. In the future, we could take a polymorphic approach where we have a RequestParser that parses a Request into an EchoRequest or a StaticFileRequest. Then, our RequestRouter uses polymorphism to automatically call the right handler. It'd be more modularity and less duplication, at the cost of more complexity.

File: lightning_server.cc [hunk 13]
```diff
@@ -1,20 +1,19 @@
 #include "config_parser.h"
 #include "lightning_server.h"
 #include "request_handlers.h"
+#include "request_router.h"
 #include "server_config.h"
-#include "request_handlers.h"
 
 #include <iostream>
+#include <cstddef>
```
Comment (KyCodeHuynh) [inline lightning_server.cc]: This gives the definition for `size_t`.

File: request_handlers.cc [hunk 14]
```diff
@@ -1,6 +1,11 @@
 #include "lightning_server.h"
 #include "request_handlers.h"
+#include "mime_types.h"
 
+#include <boost/filesystem.hpp>
+#include <boost/tokenizer.hpp>
+#include <cstddef>
```
Comment (KyCodeHuynh) [inline request_handlers.cc]: [size_t](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/42/files#r100148235)

**Comment by KyCodeHuynh** [comment]
NOTE: Failing tests are fixed as part of PR #43.

File: lightning_server.cc [hunk 15]
```diff
@@ -61,15 +60,19 @@ void LightningServer::start() {
...
-                                       response_buffer_size);
+    // TODO: Call request-router to call the right handler
+    RequestRouter router;
+    router.routeRequest(server_config_,
+                        request_buffer,
+                        request_buffer_size,
+                        response_buffer,
+                        response_buffer_size);
 
     // TODO: Use-after-free vulnerability if response_buffer is used after
```
Comment (KyCodeHuynh) [inline lightning_server.cc]: Yes, during refactoring week. We need an RAII implementation of some sort.

**Pull request merged**